### PR TITLE
Fix/#38/#39

### PR DESCRIPTION
**Pull request by mlindauer**
Fixing #38 and #39.
@mfeurer , @KEggensperger  please review

### CODE REVIEW

**Comment by KEggensperger** [comment]
When running rosenbrock I still don't get the true initial incumbent performance (default performance=120048.00). I think the first two entries should look like that:

```
{"evaluations": 0, "cost": 999999999, "cpu_time": 0, "total_cpu_time": null, "wallclock_time": 0, "incumbent": ["x1='5.0'", "x2='5'", "x3='5'", "x4='5'"]}
{"evaluations": 1, "cost":120048.00 , "cpu_time": X, "total_cpu_time": null, "wallclock_time": X, "incumbent": ["x1='5.0'", "x2='5'", "x3='5'", "x4='5'"]}
```

I would insert the first entry when initializing SMBO and the second after the initial design. What do you think?

File: smac/smbo/smbo.py [hunk 1]
```diff
@@ -120,6 +120,8 @@ def __init__(self, scenario, tae_runner=None, acquisition_function=None,
 
         self.scenario = scenario
         self.config_space = scenario.cs
+        self.traj_logger = TrajLogger(
+            output_dir=self.scenario.output_dir, stats=self.stats)
```
Comment (KEggensperger) [inline smac/smbo/smbo.py]: Do the first entry to the trajectory file here

**Comment by KEggensperger** [comment]
This PR does not fix #39, right?

**Comment by mlindauer** [comment]
Hi,

if I'm running the rosenbrock example, I get the following output in `traj_aclib2.json`

```
{"cost": 120048.0, "incumbent": ["x1='5.0'", "x2='5'", "x3='5'", "x4='5'"], "evaluations": 1, "cpu_time": 0.0015752315521240234, "total_cpu_time": null, "wallclock_time": 0.01628422737121582}
{"cost": 69492.73909731719, "incumbent": ["x1='-4.656907303634218'", "x2='-3'", "x3='2'", "x4='-2'"], "evaluations": 2, "cpu_time": 0.003468036651611328, "total_cpu_time": null, "wallclock_time": 0.20449256896972656}
```

Why do you still get the erroneous output?

> I would insert the first entry when initializing SMBO

No. Maybe we want to use an initial design that does not start with the default configuration. So, the first entry should be in the initial design.

**Comment by mlindauer** [comment]
> This PR does not fix #39, right?

Yes, it does.

**Comment by KEggensperger** [comment]
Yes you are right.
But now there is no entry at time zero, so we don't have an incumbent in case the first configuration (during the initial design) runs longer than the threshold?
@mfeurer: Is that what we want? SMAC2 works differently, doesn't it?

**Comment by mfeurer** [comment]
1. we should decide whether we add something to the trajectory at timestep zero. It would make sense to only add a configuration which actually ran. As @KEggensperger mentioned, this would break backwards compability
2. Did the original SMAC update the trajectory every time it ran the incumbent on a new instance? If yes, there should be a new entry to the trajectory in line 172.

**Comment by mlindauer** [comment]
> As @KEggensperger mentioned, this would break backwards compability

I don't see how this breaks backwards compability. smac-validate still works on the trajectory files of SMAC3. Could you please explain it.

> Did the original SMAC update the trajectory every time it ran the incumbent on a new instance? 

I don't know whether SMAC2 updates the trajectory file after each new incumbent run, but indeed, the same incumbent is listed more than once in the SMAC2's trajectory files with updated performance values.
The question is whether we also want to have the same behavior in SMAC3. As above mentioned, I don't see how the new implementation would break anything.
Furthermore, I like to avoid too much I/O accesses if we have really fast algorithm runs.
Last but not least, if we have no (or only one) instance, the estimated performance is already accurate after one run. If we have real algorithm configuration with a lot of instances, the estimated training performance is hard to use and not comparable with other SMAC runs, because each SMAC run used another instance subset to estimate the training performance.

**Comment by mlindauer** [comment]
No replies?

**Comment by mfeurer** [comment]
Yes, the workers would still run. But wouldn't the semantics of the trajectory file change? If there is no reason to change the semantics of the trajectory file, I would SMAC3 to behave as similar to SMAC2 as possible.

Regarding IO. In a scenario with short runs, how many instances do you normally have and how long does a short run take? Also, how often does the incumbent usually change?

**Comment by mlindauer** [comment]
> But wouldn't the semantics of the trajectory file change?

Is their a defined semantic of the trajectory file?
We use it normally only to get the incumbent at time step t
since the estimated training performance is incomparable anyway.
Do you use the estimated training performance in the trajectory file?

> Regarding IO. In a scenario with short runs, how many instances do you normally have and how long does a short run take?

There is no normal case for us. But as an example Spear on SWGCP: 1000 instances (plus spear is a  stochastic algorithm such that we can multiply it by infinite number of random seeds). A short run can be less than 0.1 seconds.

Cheers,
Marius

**Comment by mfeurer** [comment]
The only reference I know is a preliminary ACLIB specification, which reads like one should update every time the estimated performance changes:
    
    The configurator has to output a trajectory file in the following format:
    <cpu time used>, <wallclock time>, <number of evaluations>, <estimated performance>, <incumbent configuration>

**Comment by mlindauer** [comment]
It's funny that you argue with the AClib specification I wrote.
I don't know which version you have. But at least the current version has a footnote.

`The configurator has to output a trajectory file\footnote{A trajectory file consists of one line per new incumbent found by the configurator.} in the following JSON format`

There is later a further note what to do at the end of the configuration process:

`TIP: At the end of configuration, an entry in the trajectory file should be outputted past or at the limit of configuration, so that it is clear from examination of the file that the procedure completed successfully (e.g., there is an entry where the cpu time limit exceeded the requested limit).`

**Comment by mfeurer** [comment]
It's the only definition of trajectory I know of. I'm referring to a draft from 27.10.2014. There is no newer specification on the ACLIB website, so I assumed this is still the current version.

**Comment by mlindauer** [comment]
There was never an official/final version of this specification and even SMAC2 does not fully implements it. But here is an internal link:

https://bitbucket.org/fawcettc/aclib/src/4012624be3f445910f2779d4ecc6417b468ab00a/ac_format/?at=aclib-2.0

**Comment by mfeurer** [comment]
Okay, I will have a look at this next week.

**Comment by mfeurer** [comment]
Okay, that sound very reasonable and I'm okay with this. The only question for me is now whether we want to assume that the first entry is at time 0.0, or whether we want to change this. Since we may have different initial designs is the future, how about dropping this? @KEggensperger would this break anything on your side?

**Comment by KEggensperger** [comment]
Approved

**Comment by mlindauer** [comment]
Hi,

as discussed today, I added an entry at the beginning of the initial design such that we have something without any runs. The time spent for  the algorithms is 0 at that moment, however the wallclock time of SMAC is of course not exactly 0 (but very small).

Can we merge now?

Best,
Marius

**Comment by mfeurer** [comment]
Approved

**Pull request merged**